mono/metadata/class.c

```diff
@@ -1659,62 +1678,16 @@ mono_method_get_vtable_index (MonoMethod *method)
  *
  *   Return whenever KLASS has a finalizer, initializing klass->has_finalizer in the
  * process.
+ *
+ * LOCKING: Acquires the loader lock;
  */
 gboolean
 mono_class_has_finalizer (MonoClass *klass)
 {
-	gboolean has_finalize = FALSE;
-
-	if (klass->has_finalize_inited)
-		return klass->has_finalize;
-
-	/* Interfaces and valuetypes are not supposed to have finalizers */
-	if (!(MONO_CLASS_IS_INTERFACE (klass) || klass->valuetype)) {
-		MonoMethod *cmethod = NULL;
-
-		if (klass->rank == 1 && klass->byval_arg.type == MONO_TYPE_SZARRAY) {
-		} else if (mono_class_is_ginst (klass)) {
-			MonoClass *gklass = mono_class_get_generic_class (klass)->container_class;
-
-			has_finalize = mono_class_has_finalizer (gklass);
-		} else if (klass->parent && klass->parent->has_finalize) {
-			has_finalize = TRUE;
-		} else {
-			if (klass->parent) {
-				/*
-				 * Can't search in metadata for a method named Finalize, because that
-				 * ignores overrides.
-				 */
-				mono_class_setup_vtable (klass);
-				if (mono_class_has_failure (klass))
-					cmethod = NULL;
-				else
-					cmethod = klass->vtable [mono_class_get_object_finalize_slot ()];
-			}
-
-			if (cmethod) {
-				g_assert (klass->vtable_size > mono_class_get_object_finalize_slot ());
-
-				if (klass->parent) {
-					if (cmethod->is_inflated)
-						cmethod = ((MonoMethodInflated*)cmethod)->declaring;
-					if (cmethod != mono_class_get_default_finalize_method ())
-						has_finalize = TRUE;
-				}
-			}
-		}
-	}
-
-	mono_loader_lock ();
-	if (!klass->has_finalize_inited) {
-		klass->has_finalize = has_finalize ? 1 : 0;
-
-		mono_memory_barrier ();
-		klass->has_finalize_inited = TRUE;
-	}
-	mono_loader_unlock ();
+	if (!m_class_is_has_finalize_inited (klass))
+		mono_class_setup_has_finalizer (klass);
```

Comment:
We should probably always call `mono_class_setup_has_finalizer` (or the equivalent with the appropriate level), since already has a fast-exit check.